internal/sourcemap/metadata.go

```diff
@@ -0,0 +1,409 @@
+	logger           *logp.Logger
+	index            string
+	init             chan struct{}
+	invalidationChan chan<- Identifier
+}
+
+func NewMetadataCachingFetcher(
+	c *elasticsearch.Client,
+	backend Fetcher,
+	index string,
+	invalidationChan chan<- Identifier,
+) *MetadataCachingFetcher {
+	return &MetadataCachingFetcher{
+		esClient:         c,
+		index:            index,
+		set:              make(map[Identifier]string),
+		alias:            make(map[Identifier]*Identifier),
+		backend:          backend,
+		logger:           logp.NewLogger(logs.Sourcemap),
+		init:             make(chan struct{}),
+		invalidationChan: invalidationChan,
+	}
+}
+
+func (s *MetadataCachingFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
+	original := Identifier{name: name, version: version, path: path}
+
+	select {
+	case <-s.init:
+		// the mutex is shared by the update goroutine, we need to release it
+		// as soon as possible to avoid blocking updates.
+		if s.hasID(original) {
+			// Only fetch from ES if the sourcemap id exists
+			return s.fetch(ctx, &original)
+		}
+	default:
+		s.logger.Debugf("Metadata cache not populated. Falling back to backend fetcher for id: %s, %s, %s", name, version, path)
+		// init is in progress, ignore the metadata cache and fetch the sourcemap directly
+		// return if we get a valid sourcemap or an error
+		if c, err := s.backend.Fetch(ctx, original.name, original.version, original.path); c != nil || err != nil {
+			return c, err
+		}
+
+		s.logger.Debug("Blocking until init is completed")
+
+		// Aliases are only available after init is completed.
+		<-s.init
+	}
+
+	// path is missing from the metadata cache (and ES).
+	// Is it an alias ?
+	// Try to retrieve the sourcemap from the alias map
+	// Only fetch from ES if the sourcemap alias exists
+	if id, found := s.getAlias(original); found {
+		return s.fetch(ctx, id)
+	}
+
+	if urlPath, err := url.Parse(path); err == nil {
+		// The sourcemap coule be stored in ES with a relative
+		// bundle filepath but the request came in with an
+		// absolute path
+		original.path = urlPath.Path
+		if urlPath.Path != path && s.hasID(original) {
+			return s.fetch(ctx, &original)
+		}
+
+		// The sourcemap could be stored on ES under a certain host
+		// but a request came in from a different host.
+		// Look for an alias to the url path to retrieve the correct
+		// host and fetch the sourcemap
+		if id, found := s.getAlias(original); found {
+			return s.fetch(ctx, id)
+		}
+
+		// Clean the url and try again if the result is different from
+		// the original bundle filepath
+		urlPath.RawQuery = ""
+		urlPath.Fragment = ""
+		urlPath = urlPath.JoinPath()
+		cleanPath := urlPath.String()
+
+		if cleanPath != path {
+			s.logger.Debugf("original filepath %s converted to %s", path, cleanPath)
+			return s.Fetch(ctx, name, version, cleanPath)
+		}
+	}
+
+	return nil, nil
+}
+
+func (s *MetadataCachingFetcher) hasID(key Identifier) bool {
```

Comment:
How about abstracting the concept of `aliases` to the caller and changing this to `func (s *MetadataCachingFetcher) getID(key Identifier) string`. If the key is not presented, return an empty string, otherwise return the ID, even if the `key` is an alias. This way the caller would always only have to make one call and not be concerned with these internas.

Reply:
Done! :+1: